lighthouse-core/lib/client-rect-functions.js

```diff
@@ -0,0 +1,311 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * @param {LH.Artifacts.ClientRect} cr
+ * @param {{x:number, y:number}} point
+ */
+/* istanbul ignore next */
+function rectContainsPoint(cr, {x, y}) {
+  return cr.left <= x && cr.right >= x && cr.top <= y && cr.bottom >= y;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} cr1
+ * @param {LH.Artifacts.ClientRect} cr2
+ */
+/* istanbul ignore next */
+function rectContains(cr1, cr2) {
+  const topLeft = {
+    x: cr2.left,
+    y: cr2.top,
+  };
+  const topRight = {
+    x: cr2.right,
+    y: cr2.top,
+  };
+  const bottomLeft = {
+    x: cr2.left,
+    y: cr2.bottom,
+  };
+  const bottomRight = {
+    x: cr2.right,
+    y: cr2.bottom,
+  };
+  return (
+    rectContainsPoint(cr1, topLeft) &&
+    rectContainsPoint(cr1, topRight) &&
+    rectContainsPoint(cr1, bottomLeft) &&
+    rectContainsPoint(cr1, bottomRight)
+  );
+}
+
+/**
+ * Merge client rects together. This may result in a larger overall size than that of the individual client rects.
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ */
+function simplifyClientRects(clientRects) {
+  clientRects = filterOutTinyClientRects(clientRects);
+  clientRects = filterOutClientRectsContainedByOthers(clientRects);
+  clientRects = mergeTouchingClientRects(clientRects);
+  return clientRects;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function filterOutTinyClientRects(clientRects) {
+  // 1x1px rect shouldn't be reason to treat the rect as something the user should tap on.
```

Comment:
A tad unfortunate that we're letting the tap targets bit leak out into here :/ 

maybe we can clarify a bit in simplifyClientRects that it's removing small ones?

Reply:
> A tad unfortunate that we're letting the tap targets bit leak out into here :/

True... some of the logic isn't really useful in a more general way though, especially the `simplifyClientRects` logic is very specific to this audit. 

Maybe we can split it into two files, `rectHelpers` and `tapTargetClientRectHelpers`? And in the `rectHelpers` we can address @Hoten's point about renaming to just `rects` then.

Reply:
I've thought about this a bit more, and I'm actually somewhat convinced that this is something we'd probably want to do with anything that's looking at real objects on a page, so it's really just the "tap" language I think I'm pushing back a bit on. 

I'm fine either way with A) just documenting in the jsdoc what exactly it does to the rects during simplification or B) splitting into two files as you suggested